src/people.rs

```diff
@@ -1204,15 +1207,52 @@ impl ContextPeopleExt for Context {
         );
     }
 
-    fn sample_person<R: RngId + 'static>(&self, rng_id: R) -> Result<PersonId, IxaError>
+    #[allow(clippy::cast_sign_loss, clippy::cast_possible_truncation)]
+    fn sample_person<R: RngId + 'static, T: Query>(
+        &self,
+        rng_id: R,
+        q: T,
+    ) -> Result<PersonId, IxaError>
     where
         R::RngType: Rng,
     {
         if self.get_current_population() == 0 {
             return Err(IxaError::IxaError(String::from("Empty population")));
         }
-        let result = self.sample_range(rng_id, 0..self.get_current_population());
-        Ok(PersonId { id: result })
+
+        // Special case the empty query because we can do it in O(1).
+        if q.get_query().is_empty() {
+            let result = self.sample_range(rng_id, 0..self.get_current_population());
+            return Ok(PersonId { id: result });
+        }
+
+        T::setup(self);
+
+        // This function implements "Algorithm L" from KIM-HUNG LI
+        // Reservoir-Sampling Algorithms of Time Complexity O(n(1 + log(N/n)))
+        // https://dl.acm.org/doi/pdf/10.1145/198429.198435
+        // Temporary variables.
+        let mut selected: Option<PersonId> = None;
+        let mut w: f64 = self.sample_range(rng_id, 0.0..1.0);
+        let mut ctr: usize = 0;
+        let mut i: usize = 1;
+
+        self.query_people_internal(
+            |person| {
+                ctr += 1;
+                if i == ctr {
+                    selected = Some(person);
+                    i += (f64::ln(self.sample_range(rng_id, 0.0..1.0)) / f64::ln(1.0 - w)).floor()
+                        as usize
+                        + 1;
+                    w *= self.sample_range(rng_id, 0.0..1.0);
+                }
+            },
+            q.get_query(),
+        );
+
+        // This should always be set by the above algorithm.
+        Ok(selected.unwrap())
```

Comment:
This actually needs to check for None and then generate an error rather than just crashing.